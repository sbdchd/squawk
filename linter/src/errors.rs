use squawk_parser::error::PGQueryError;

#[derive(Debug, PartialEq)]
pub enum CheckSQLError {
    ParsingSQL(PGQueryError),
}

impl std::convert::From<PGQueryError> for CheckSQLError {
    fn from(err: PGQueryError) -> Self {
        Self::ParsingSQL(err)
    }
}
