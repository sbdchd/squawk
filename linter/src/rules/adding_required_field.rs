use crate::versions::Version;
use crate::violations::{RuleViolation, RuleViolationKind};

use squawk_parser::ast::{
    AlterTableCmds, AlterTableDef, AlterTableType, ColumnDefConstraint, ConstrType, RawStmt, Stmt,
};

fn has_not_null_and_no_default_constraint(constraints: &[ColumnDefConstraint]) -> bool {
    let mut has_not_null = false;
    let mut has_default = false;
    for ColumnDefConstraint::Constraint(constraint) in constraints {
        if constraint.contype == ConstrType::NotNull {
            has_not_null = true;
        }
        if constraint.contype == ConstrType::Default {
            has_default = true;
        }
    }
    has_not_null && !has_default
}

#[must_use]
pub fn adding_required_field(
    tree: &[RawStmt],
    _pg_version: Option<Version>,
    _assume_in_transaction: bool,
) -> Vec<RuleViolation> {
    let mut errs = vec![];

    for raw_stmt in tree {
        match &raw_stmt.stmt {
            Stmt::AlterTableStmt(stmt) => {
                for AlterTableCmds::AlterTableCmd(cmd) in &stmt.cmds {
                    if cmd.subtype == AlterTableType::AddColumn {
                        if let Some(AlterTableDef::ColumnDef(column_def)) = &cmd.def {
                            if has_not_null_and_no_default_constraint(&column_def.constraints) {
                                errs.push(RuleViolation::new(
                                    RuleViolationKind::AddingRequiredField,
                                    raw_stmt.into(),
                                    None,
                                ));
                            }
                        }
                    }
                }
            }
            _ => continue,
        }
    }

    errs
}

#[cfg(test)]
mod test_rules {
    use crate::{
        check_sql_with_rule,
        violations::{RuleViolation, RuleViolationKind},
    };
    use insta::assert_debug_snapshot;

    fn lint_sql(sql: &str) -> Vec<RuleViolation> {
        check_sql_with_rule(sql, &RuleViolationKind::AddingRequiredField, None, false).unwrap()
    }

    #[test]
    fn test_nullable() {
        let ok_sql = r#"
ALTER TABLE "recipe" ADD COLUMN "public" boolean;
  "#;
        assert_debug_snapshot!(lint_sql(ok_sql));
    }

    #[test]
    fn test_not_null_with_default() {
        let ok_sql = r#"
ALTER TABLE "recipe" ADD COLUMN "public" boolean NOT NULL DEFAULT true;
  "#;
        assert_debug_snapshot!(lint_sql(ok_sql));
    }

    #[test]
    fn test_not_null_without_default() {
        let bad_sql = r#"
ALTER TABLE "recipe" ADD COLUMN "public" boolean NOT NULL;
  "#;
        assert_debug_snapshot!(lint_sql(bad_sql));
    }
}
