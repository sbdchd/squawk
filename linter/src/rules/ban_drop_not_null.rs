use crate::{versions::Version, violations::RuleViolation, violations::RuleViolationKind};

use squawk_parser::ast::{AlterTableCmds, AlterTableType, RawStmt, Stmt};

#[must_use]
pub fn ban_drop_not_null(
    tree: &[RawStmt],
    _pg_version: Option<Version>,
    _assume_in_transaction: bool,
) -> Vec<RuleViolation> {
    let mut errs = vec![];
    for raw_stmt in tree {
        match &raw_stmt.stmt {
            Stmt::AlterTableStmt(stmt) => {
                for cmd in &stmt.cmds {
                    let AlterTableCmds::AlterTableCmd(cmd) = cmd;
                    if cmd.subtype == AlterTableType::DropNotNull {
                        errs.push(RuleViolation::new(
                            RuleViolationKind::BanDropNotNull,
                            raw_stmt.into(),
                            None,
                        ));
                    }
                }
            }
            _ => continue,
        }
    }
    errs
}

#[cfg(test)]
mod test_rules {
    use crate::{
        check_sql_with_rule,
        violations::{RuleViolation, RuleViolationKind},
    };
    use insta::assert_debug_snapshot;

    fn lint_sql(sql: &str) -> Vec<RuleViolation> {
        check_sql_with_rule(sql, &RuleViolationKind::BanDropNotNull, None, false).unwrap()
    }

    #[test]
    fn test_ban_drop_not_null() {
        let bad_sql = r#"
ALTER TABLE "bar_tbl" ALTER COLUMN "foo_col" DROP NOT NULL;
  "#;
        assert_debug_snapshot!(lint_sql(bad_sql));
    }
}
