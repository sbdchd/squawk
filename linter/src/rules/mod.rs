pub mod adding_field_with_default;
mod utils;
pub use adding_field_with_default::*;
pub mod adding_not_null_field;
pub use adding_not_null_field::*;
pub mod adding_primary_key_constraint;
pub use adding_primary_key_constraint::*;
pub mod bad_drop_database;
pub use bad_drop_database::*;
pub mod changing_column_type;
pub use changing_column_type::*;
pub mod constraint_missing_not_valid;
pub use constraint_missing_not_valid::*;
pub mod disallow_unique_constraint;
pub use disallow_unique_constraint::*;
pub mod prefer_robust_stmts;
pub use prefer_robust_stmts::*;
pub mod prefer_text_field;
pub use prefer_text_field::*;
pub mod renaming_column;
pub use renaming_column::*;
pub mod renaming_table;
pub use renaming_table::*;
pub mod ban_drop_column;
pub use ban_drop_column::*;
pub mod require_concurrent_index_creation;
pub use require_concurrent_index_creation::*;
pub mod ban_char_field;
pub mod require_concurrent_index_deletion;
pub use ban_char_field::*;
pub use require_concurrent_index_deletion::*;
pub mod adding_foreign_key_constraint;
pub use adding_foreign_key_constraint::*;
pub mod prefer_timestamptz;
pub use prefer_timestamptz::*;
