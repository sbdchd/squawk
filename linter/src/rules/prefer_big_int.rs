use std::collections::HashSet;

use crate::{
    versions::Version,
    violations::{RuleViolation, RuleViolationKind},
};

use squawk_parser::ast::{ColumnDef, RawStmt};

use super::utils::columns_create_or_modified;

#[must_use]
pub fn prefer_big_int(
    tree: &[RawStmt],
    _pg_version: Option<Version>,
    _assume_in_transaction: bool,
) -> Vec<RuleViolation> {
    let mut errs = vec![];
    for raw_stmt in tree {
        for column in columns_create_or_modified(&raw_stmt.stmt) {
            check_column_def(&mut errs, raw_stmt, column);
        }
    }
    errs
}

lazy_static! {
    static ref SMALL_INT_TYPES: HashSet<&'static str> = HashSet::from([
        "smallint",
        "integer",
        "int2",
        "int4",
        "serial",
        "serial2",
        "serial4",
        "smallserial",
    ]);
}

fn check_column_def(errs: &mut Vec<RuleViolation>, raw_stmt: &RawStmt, column_def: &ColumnDef) {
    if let Some(column_name) = column_def.type_name.names.last() {
        if SMALL_INT_TYPES.contains(column_name.string.sval.as_str()) {
            errs.push(RuleViolation::new(
                RuleViolationKind::PreferBigInt,
                raw_stmt.into(),
                None,
            ));
        }
    }
}

#[cfg(test)]
mod test_rules {
    use insta::assert_debug_snapshot;

    use crate::{
        check_sql_with_rule,
        rules::test_utils::violations_to_kinds,
        violations::{RuleViolation, RuleViolationKind},
    };

    fn lint_sql(sql: &str) -> Vec<RuleViolation> {
        check_sql_with_rule(sql, &RuleViolationKind::PreferBigInt, None, false).unwrap()
    }

    #[test]
    fn test_create_table_ok() {
        let ok_sql = r"
create table users (
    id bigint
);
create table users (
    id int8
);
create table users (
    id bigserial
);
create table users (
    id serial8
);
  ";
        assert_eq!(lint_sql(ok_sql), vec![]);
    }
    #[test]
    fn test_create_table_bad() {
        let bad_sql = r"
create table users (
    id smallint
);
create table users (
    id int2
);
create table users (
    id integer
);
create table users (
    id int4
);
create table users (
    id serial
);
create table users (
    id serial2
);
create table users (
    id serial4
);
create table users (
    id smallserial
);
  ";
        let res = lint_sql(bad_sql);
        let violations = violations_to_kinds(&res);
        assert_eq!(
            violations.len(),
            8,
            "we should have 8 statements with violations"
        );
        assert_eq!(
            violations.len(),
            violations
                .into_iter()
                .filter(|v| { *v == RuleViolationKind::PreferBigInt })
                .count(),
            "all violations should be big int violations"
        );
        assert_debug_snapshot!(res);
    }
}
