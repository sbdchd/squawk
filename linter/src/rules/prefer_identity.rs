use std::collections::HashSet;

use crate::{
    versions::Version,
    violations::{RuleViolation, RuleViolationKind},
};

use squawk_parser::ast::{ColumnDef, RawStmt};

use super::utils::columns_create_or_modified;

#[must_use]
pub fn prefer_identity(
    tree: &[RawStmt],
    _pg_version: Option<Version>,
    _assume_in_transaction: bool,
) -> Vec<RuleViolation> {
    let mut errs = vec![];
    for raw_stmt in tree {
        for column in columns_create_or_modified(&raw_stmt.stmt) {
            check_column_def(&mut errs, raw_stmt, column);
        }
    }
    errs
}

lazy_static! {
    static ref SERIAL_TYPES: HashSet<&'static str> = HashSet::from([
        "serial",
        "serial2",
        "serial4",
        "serial8",
        "smallserial",
        "bigserial",
    ]);
}

fn check_column_def(errs: &mut Vec<RuleViolation>, raw_stmt: &RawStmt, column_def: &ColumnDef) {
    if let Some(column_name) = column_def.type_name.names.last() {
        if SERIAL_TYPES.contains(column_name.string.sval.as_str()) {
            errs.push(RuleViolation::new(
                RuleViolationKind::PreferIdentity,
                raw_stmt.into(),
                None,
            ));
        }
    }
}

#[cfg(test)]
mod test_rules {
    use insta::assert_debug_snapshot;

    use crate::{
        check_sql_with_rule,
        rules::test_utils::violations_to_kinds,
        violations::{RuleViolation, RuleViolationKind},
    };
    fn lint_sql(sql: &str) -> Vec<RuleViolation> {
        check_sql_with_rule(sql, &RuleViolationKind::PreferIdentity, None, false).unwrap()
    }

    #[test]
    fn test_prefer_identity_bad() {
        let bad_sql = r"
create table users (
    id serial
);
create table users (
    id serial2
);
create table users (
    id serial4
);
create table users (
    id serial8
);
create table users (
    id smallserial
);
create table users (
    id bigserial
);
  ";

        let res = lint_sql(bad_sql);
        let violations = violations_to_kinds(&res);
        assert_eq!(
            violations.len(),
            6,
            "we should have 6 statements with violations"
        );
        assert_eq!(
            violations.len(),
            violations
                .into_iter()
                .filter(|v| { *v == RuleViolationKind::PreferIdentity })
                .count(),
            "all violations should be prefer-identity violations"
        );

        assert_debug_snapshot!(res);
    }
    #[test]
    fn test_prefer_identity_ok() {
        let ok_sql = r"
create table users (
    id  bigint generated by default as identity primary key
);
create table users (
    id  bigint generated always as identity primary key
);
  ";

        assert_eq!(lint_sql(ok_sql), vec![]);
    }
}
