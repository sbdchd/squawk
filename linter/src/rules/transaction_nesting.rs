use crate::versions::Version;
use crate::violations::{RuleViolation, RuleViolationKind};
use crate::ViolationMessage;

use squawk_parser::ast::{RawStmt, Stmt, TransactionStmtKind};

#[must_use]
pub fn transaction_nesting(
    tree: &[RawStmt],
    _pg_version: Option<Version>,
    assume_in_transaction: bool,
) -> Vec<RuleViolation> {
    let mut errs = vec![];
    let mut in_explicit_transaction = false;
    let assume_in_transaction_help = ViolationMessage::Help(
        "Put migration statements in separate files to have them be in separate transactions or \
        don't use the assume-in-transaction setting."
            .into(),
    );

    for raw_stmt in tree {
        match &raw_stmt.stmt {
            Stmt::TransactionStmt(stmt) => match stmt.kind {
                TransactionStmtKind::Begin | TransactionStmtKind::Start => {
                    if assume_in_transaction {
                        errs.push(RuleViolation::new(
                                RuleViolationKind::TransactionNesting,
                                raw_stmt.into(),
                                Some(vec![
                                    ViolationMessage::Note(
                                        "There is an existing transaction already in progress, managed by your migration tool.".into()
                                    ),
                                    assume_in_transaction_help.clone(),
                                ])
                            ));
                    } else if in_explicit_transaction {
                        errs.push(RuleViolation::new(
                            RuleViolationKind::TransactionNesting,
                            raw_stmt.into(),
                            None,
                        ));
                    }
                    in_explicit_transaction = true;
                }
                TransactionStmtKind::Commit | TransactionStmtKind::Rollback => {
                    if assume_in_transaction {
                        errs.push(RuleViolation::new(
                                RuleViolationKind::TransactionNesting,
                                raw_stmt.into(),
                                Some(vec![
                                    ViolationMessage::Note(
                                        "Attempting to end the transaction that is managed by your migration tool.".into()
                                    ),
                                    assume_in_transaction_help.clone(),
                                ])
                            ));
                    } else if !in_explicit_transaction {
                        errs.push(RuleViolation::new(
                                RuleViolationKind::TransactionNesting,
                                raw_stmt.into(),
                                Some(vec![
                                    ViolationMessage::Note(
                                        "There is no transaction to COMMIT or ROLLBACK.".into()
                                    ),
                                    ViolationMessage::Help(
                                        "BEGIN a transaction at an earlier point in the migration or remove this statement.".into()
                                    ),
                                ])
                            ));
                    }
                    in_explicit_transaction = false;
                }
                _ => continue,
            },
            _ => continue,
        }
    }
    errs
}

#[cfg(test)]
mod test_rules {
    use crate::{
        check_sql_with_rule,
        violations::{RuleViolation, RuleViolationKind},
    };
    use insta::assert_debug_snapshot;

    fn lint_sql(sql: &str) -> Vec<RuleViolation> {
        check_sql_with_rule(sql, &RuleViolationKind::TransactionNesting, None, false).unwrap()
    }
    fn lint_sql_assuming_in_transaction(sql: &str) -> Vec<RuleViolation> {
        check_sql_with_rule(sql, &RuleViolationKind::TransactionNesting, None, true).unwrap()
    }

    #[test]
    fn test_no_nesting() {
        let ok_sql = r"
BEGIN;
SELECT 1;
COMMIT;
  ";
        assert_eq!(lint_sql(ok_sql), vec![]);
    }

    #[test]
    fn test_no_nesting_repeated() {
        let ok_sql = r"
BEGIN;
SELECT 1;
COMMIT;
-- This probably shouldn't be done in a migration. However, Squawk may be linting several
-- migrations that are concatentated, so don't raise a warning here.
BEGIN;
SELECT 2;
COMMIT;
  ";
        assert_eq!(lint_sql(ok_sql), vec![]);
    }

    #[test]
    fn test_no_nesting_with_assume_in_transaction() {
        let ok_sql = r"
SELECT 1;
  ";
        assert_eq!(lint_sql_assuming_in_transaction(ok_sql), vec![]);
    }

    #[test]
    fn test_begin_repeated() {
        let bad_sql = r"
BEGIN;
BEGIN;
SELECT 1;
COMMIT;
  ";
        assert_debug_snapshot!(lint_sql(bad_sql));
    }

    #[test]
    fn test_begin_with_assume_in_transaction() {
        let bad_sql = r"
BEGIN;
SELECT 1;
  ";
        assert_debug_snapshot!(lint_sql_assuming_in_transaction(bad_sql));
    }

    #[test]
    fn test_commit_repeated() {
        let bad_sql = r"
BEGIN;
SELECT 1;
COMMIT;
COMMIT;
  ";
        assert_debug_snapshot!(lint_sql(bad_sql));
    }

    #[test]
    fn test_commit_with_assume_in_transaction() {
        let bad_sql = r"
SELECT 1;
COMMIT;
  ";
        assert_debug_snapshot!(lint_sql_assuming_in_transaction(bad_sql));
    }

    #[test]
    fn test_rollback_with_assume_in_transaction() {
        let bad_sql = r"
SELECT 1;
-- Not sure why rollback would be used in a migration, but test for completeness
ROLLBACK;
  ";
        assert_debug_snapshot!(lint_sql_assuming_in_transaction(bad_sql));
    }
}
