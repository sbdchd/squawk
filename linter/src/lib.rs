#![allow(clippy::shadow_unrelated)]
#![allow(clippy::missing_errors_doc)]
pub mod errors;
pub mod rules;
pub mod versions;
pub mod violations;
#[macro_use]
extern crate lazy_static;

use crate::errors::CheckSqlError;
use crate::rules::adding_required_field;
use crate::rules::ban_drop_not_null;
use crate::rules::prefer_big_int;
use crate::rules::prefer_identity;
use crate::rules::transaction_nesting;
use crate::rules::{
    adding_field_with_default, adding_foreign_key_constraint, adding_not_nullable_field,
    adding_primary_key_constraint, ban_char_type, ban_drop_column, ban_drop_database,
    ban_drop_table, changing_column_type, constraint_missing_not_valid, disallow_unique_constraint,
    prefer_bigint_over_int, prefer_bigint_over_smallint, prefer_robust_stmts, prefer_text_field,
    prefer_timestamptz, renaming_column, renaming_table, require_concurrent_index_creation,
    require_concurrent_index_deletion,
};
use crate::violations::{RuleViolation, RuleViolationKind, ViolationMessage};
use squawk_parser::ast::RawStmt;
use squawk_parser::parse::parse_sql_query;
use std::collections::HashSet;
use versions::Version;

#[derive(Clone)]
pub struct SquawkRule {
    pub name: RuleViolationKind,
    func: fn(&[RawStmt], Option<Version>, bool) -> Vec<RuleViolation>,
    pub messages: Vec<ViolationMessage>,
}

lazy_static! {
    pub static ref RULES: Vec<SquawkRule> = vec![
    // see ChangingColumnType
    SquawkRule {
        name: RuleViolationKind::AddingFieldWithDefault,
        func: adding_field_with_default,
        messages: vec![
            ViolationMessage::Note(
                "Adding a field with a VOLATILE DEFAULT requires a table table rewrite with an ACCESS EXCLUSIVE lock. In Postgres versions 11+, non-VOLATILE DEFAULTs can be added without a rewrite.".into(),
            ),
            ViolationMessage::Help(
                "Add the field as nullable, then set a default, backfill, and remove nullabilty.".into(),
            ),

        ],
    },
    SquawkRule {
        name: RuleViolationKind::AddingForeignKeyConstraint,
        func: adding_foreign_key_constraint,
        messages: vec![
            ViolationMessage::Note(
                "Requires a table scan of the table you're altering and a SHARE ROW EXCLUSIVE lock on both tables, which blocks writes to both tables while your table is scanned.".into()
            ),
            ViolationMessage::Help("Add NOT VALID to the constraint in one transaction and then VALIDATE the constraint in a separate transaction.".into()),
        ]
    },
    // usually paired with a DEFAULT
    SquawkRule {
        name: RuleViolationKind::AddingNotNullableField,
        func: adding_not_nullable_field,
        messages: vec![
            // https://www.postgresql.org/docs/10/sql-altertable.html
            ViolationMessage::Note(
                "Adding a NOT NULL field requires exclusive locks and table rewrites.".into(),
            ),
            ViolationMessage::Help("Make the field nullable.".into())
        ],
    },
    SquawkRule {
        name: RuleViolationKind::AddingRequiredField,
        func: adding_required_field,
        messages: vec![
            ViolationMessage::Note(
                "Adding a NOT NULL field without a DEFAULT will fail for a populated table.".into()
            ),
            ViolationMessage::Help(
                "Make the field nullable or add a non-VOLATILE DEFAULT (Postgres 11+).".into()
            ),
        ],
    },
    SquawkRule {
        name: RuleViolationKind::AddingSerialPrimaryKeyField,
        func: adding_primary_key_constraint,
        messages: vec![
            ViolationMessage::Note(
                "Adding a PRIMARY KEY constraint results in locks and table rewrites".into(),
            ),
            ViolationMessage::Help(
                "Add the PRIMARY KEY constraint USING an index.".into(),
            ),

        ],
    },
    SquawkRule {
        name: RuleViolationKind::BanCharField,
        func: ban_char_type,
        messages: vec![
            ViolationMessage::Help(
                "Use text or varchar instead.".into()
            ),
        ]
    },
    SquawkRule {
        name: RuleViolationKind::BanDropColumn,
        func: ban_drop_column,
        messages: vec![
            ViolationMessage::Note(
                "Dropping a column may break existing clients.".into()
            ),
        ],
    },
    SquawkRule {
        name: RuleViolationKind::BanDropDatabase,
        func: ban_drop_database,
        messages: vec![
            ViolationMessage::Note(
                "Dropping a database may break existing clients.".into()
            )
        ],
    },
    SquawkRule {
        name: RuleViolationKind::BanDropNotNull,
        func: ban_drop_not_null,
        messages: vec![
            ViolationMessage::Note(
                "Dropping a NOT NULL constraint may break existing clients.".into()
            ),
        ],
    },
    SquawkRule {
        name: RuleViolationKind::BanDropTable,
        func: ban_drop_table,
        messages: vec![
            ViolationMessage::Note(
                "Dropping a table may break existing clients.".into()
            )
        ],
    },
    // > Adding a column with a volatile DEFAULT or changing the type of an
    // > existing column will require the entire table and its indexes to be
    // > rewritten. As an exception, when changing the type of an existing
    // > column, if the USING clause does not change the column contents and
    // > the old type is either binary coercible to the new type or an
    // > unconstrained domain over the new type, a table rewrite is not
    // > needed; but any indexes on the affected columns must still be
    // > rebuilt. Table and/or index rebuilds may take a significant amount of
    // > time for a large table; and will temporarily require as much as
    // > double the disk space.
    // https://www.postgresql.org/docs/current/sql-altertable.html
    SquawkRule {
        name: RuleViolationKind::ChangingColumnType,
        func: changing_column_type,
        messages: vec![
            ViolationMessage::Note("Requires an ACCESS EXCLUSIVE lock on the table which blocks reads.".into()),
            ViolationMessage::Note("Changing the type may break existing clients.".into()),
        ],
    },

    // > Scanning a large table to verify a new foreign key or check
    // > constraint can take a long time, and other updates to the table are
    // > locked out until the ALTER TABLE ADD CONSTRAINT command is committed.
    // > The main purpose of the NOT VALID constraint option is to reduce the
    // > impact of adding a constraint on concurrent updates. With NOT VALID,
    // > the ADD CONSTRAINT command does not scan the table and can be
    // > committed immediately. After that, a VALIDATE CONSTRAINT command can
    // > be issued to verify that existing rows satisfy the constraint. The
    // > validation step does not need to lock out concurrent updates, since
    // > it knows that other transactions will be enforcing the constraint for
    // > rows that they insert or update; only pre-existing rows need to be
    // > checked. Hence, validation acquires only a SHARE UPDATE EXCLUSIVE
    // > lock on the table being altered. (If the constraint is a foreign key
    // > then a ROW SHARE lock is also required on the table referenced by the
    // > constraint.) In addition to improving concurrency, it can be useful
    // > to use NOT VALID and VALIDATE CONSTRAINT in cases where the table is
    // > known to contain pre-existing violations. Once the constraint is in
    // > place, no new violations can be inserted, and the existing problems
    // > can be corrected at leisure until VALIDATE CONSTRAINT finally
    // > succeeds.
    // https://www.postgresql.org/docs/current/sql-altertable.html#SQL-ALTERTABLE-NOTES
    SquawkRule {
        name: RuleViolationKind::ConstraintMissingNotValid,
        func: constraint_missing_not_valid,
        messages: vec![
            ViolationMessage::Note("Requires a table scan to verify constraint and an ACCESS EXCLUSIVE lock which blocks reads.".into()),
            ViolationMessage::Help("Add NOT VALID to the constraint in one transaction and then VALIDATE the constraint in a separate transaction.".into()),
        ],
    },
    // > Although most forms of ADD table_constraint require an ACCESS
    // > EXCLUSIVE lock, ADD FOREIGN KEY requires only a SHARE ROW EXCLUSIVE
    // > lock.
    // https://www.postgresql.org/docs/current/sql-altertable.html
    SquawkRule {
        name: RuleViolationKind::DisallowedUniqueConstraint,
        func: disallow_unique_constraint,
        messages: vec![
            ViolationMessage::Note(
                "Adding a UNIQUE constraint requires an ACCESS EXCLUSIVE lock which blocks reads.".into(),
            ),
            ViolationMessage::Help(
                "Create an index CONCURRENTLY and create the constraint using the index.".into(),
            ),

        ],
    },
    SquawkRule {
        name: RuleViolationKind::PreferBigInt,
        func: prefer_big_int,
        messages: vec![
            ViolationMessage::Note(
                "Hitting the max 32 bit integer is possible and may break your application.".into()
            ),
            ViolationMessage::Help(
                "Use 64bit integer values instead to prevent hitting this limit.".into()
            ),
        ],
    },
    SquawkRule {
        name: RuleViolationKind::PreferBigintOverInt,
        func: prefer_bigint_over_int,
        messages: vec![
            ViolationMessage::Note(
                "Hitting the max 32 bit integer is possible and may break your application.".into()
            ),
            ViolationMessage::Help(
                "Use 64bit integer values instead to prevent hitting this limit.".into()
            ),
        ],
    },
    SquawkRule {
        name: RuleViolationKind::PreferBigintOverSmallint,
        func: prefer_bigint_over_smallint,
        messages: vec![
            ViolationMessage::Note(
                "Hitting the max 16 bit integer is possible and may break your application.".into()
            ),
            ViolationMessage::Help(
                "Use 64bit integer values instead to prevent hitting this limit.".into()
            ),
        ],
    },
    SquawkRule {
        name: RuleViolationKind::PreferIdentity,
        func: prefer_identity,
        messages: vec![
            ViolationMessage::Note(
                "Serial types have confusing behaviors that make schema management difficult.".into()
            ),
            ViolationMessage::Help(
                "Use identity columns instead for more features and better usability.".into()
            ),
        ],
    },
    SquawkRule {
        name: RuleViolationKind::PreferRobustStmts,
        func: prefer_robust_stmts,
        messages: vec![
            ViolationMessage::Help(
                "Consider wrapping in a transaction or adding a IF NOT EXISTS clause if the statement supports it.".into()
            ),
        ]
    },
    // see ConstraintMissingNotValid for more docs
    SquawkRule {
        name: RuleViolationKind::PreferTextField,
        func: prefer_text_field,
        messages: vec![
            ViolationMessage::Note(
                "Changing the size of a varchar field requires an ACCESS EXCLUSIVE lock.".into()
            ),
            ViolationMessage::Help(
                "Use a text field with a check constraint.".into()
            ),
        ]
    },
    SquawkRule {
        name: RuleViolationKind::PreferTimestampTz,
        func: prefer_timestamptz,
        messages: vec![
            ViolationMessage::Note(
                "A timestamp field without a timezone can lead to data loss, depending on your database session timezone.".into()
            ),
            ViolationMessage::Help(
                "Use timestamptz instead of timestamp for your column type.".into()
            ),
        ]
    },
    // > The RENAME forms change the name of a table (or an index, sequence,
    // > view, materialized view, or foreign table), the name of an individual
    // > column in a table, or the name of a constraint of the table. There is
    // > no effect on the stored data.
    // https://www.postgresql.org/docs/10/sql-altertable.html
    SquawkRule {
        name: RuleViolationKind::RenamingColumn,
        func: renaming_column,
        messages: vec![
            ViolationMessage::Note(
                "Renaming a column may break existing clients.".into()
            ),
        ],
    },
    // see RenamingColumn rule
    SquawkRule {
        name: RuleViolationKind::RenamingTable,
        func: renaming_table,
        messages: vec![
            ViolationMessage::Note(
                "Renaming a table may break existing clients.".into()
            ),
        ],
    },
    // https://www.postgresql.org/docs/10/sql-createindex.html#SQL-CREATEINDEX-CONCURRENTLY
    SquawkRule {
        name: RuleViolationKind::RequireConcurrentIndexCreation,
        func: require_concurrent_index_creation,
        messages: vec![
            ViolationMessage::Note(
                "Creating an index blocks writes.".into()
            ),
            ViolationMessage::Help(
                "Create the index CONCURRENTLY.".into()
            ),
        ],
    },
    // https://www.postgresql.org/docs/10/sql-dropindex.html
    SquawkRule {
        name: RuleViolationKind::RequireConcurrentIndexDeletion,
        func: require_concurrent_index_deletion,
        messages: vec![
            ViolationMessage::Note(
                "Deleting an index blocks selects, inserts, updates, and deletes on the index's table.".into()
            ),
            ViolationMessage::Help(
                "Delete the index CONCURRENTLY.".into()
            ),
        ],
    },
    SquawkRule {
        name: RuleViolationKind::TransactionNesting,
        func: transaction_nesting,
        messages: vec![
            ViolationMessage::Note(
                "There is an existing transaction already in progress.".into()
            ),
            ViolationMessage::Help(
                "COMMIT the previous transaction before issuing a BEGIN or START TRANSACTION statement.".into()
            ),
        ],
    },
    // generator::new-rule-above
    ];

}

pub fn check_sql(
    sql: &str,
    excluded_rules: &[RuleViolationKind],
    pg_version: Option<Version>,
    assume_in_transaction: bool,
) -> Result<Vec<RuleViolation>, CheckSqlError> {
    let tree = parse_sql_query(sql)?;

    let excluded_rules: HashSet<RuleViolationKind> = excluded_rules.iter().cloned().collect();

    let mut errs = vec![];
    for rule in RULES.iter().filter(|r| !excluded_rules.contains(&r.name)) {
        errs.extend((rule.func)(&tree, pg_version, assume_in_transaction));
    }

    errs.sort_by_key(|v| v.span.start);

    Ok(errs)
}

pub fn check_sql_with_rule(
    sql: &str,
    rule_kind: &RuleViolationKind,
    pg_version: Option<Version>,
    assume_in_transaction: bool,
) -> Result<Vec<RuleViolation>, CheckSqlError> {
    let tree = parse_sql_query(sql)?;
    let mut errs = vec![];
    for rule in RULES.iter() {
        if rule.name == *rule_kind {
            errs.extend((rule.func)(&tree, pg_version, assume_in_transaction));
        }
    }

    errs.sort_by_key(|v| v.span.start);

    Ok(errs)
}

#[cfg(test)]
mod test_rules {
    use super::*;
    use insta::{assert_debug_snapshot, assert_display_snapshot};
    use std::convert::TryFrom;
    use std::str::FromStr;

    #[test]
    fn rules_should_be_sorted() {
        let original_rules: Vec<String> = RULES.iter().map(|x| x.name.to_string()).collect();
        let mut sorted_rule_ids = original_rules.clone();
        sorted_rule_ids.sort();
        assert_eq!(original_rules, sorted_rule_ids);
    }
    /// Ensure we handle both serializing and deserializing `RuleViolationKind`
    #[test]
    fn test_parsing_rule_kind() {
        let rule_names = RULES.iter().map(|r| r.name.clone());
        for rule in rule_names {
            let rule_str = rule.to_string();
            assert_eq!(RuleViolationKind::from_str(&rule_str), Ok(rule.clone()));
            assert_eq!(RuleViolationKind::try_from(rule_str.as_ref()), Ok(rule));
        }
    }
    /// Ensure rule names don't change
    #[test]
    fn test_rule_names_debug_snap() {
        let rule_names: Vec<String> = RULES.iter().map(|r| r.name.to_string()).collect();
        assert_debug_snapshot!(rule_names);
    }
    #[test]
    fn test_rule_names_display_snap() {
        let rule_names: Vec<String> = RULES.iter().map(|r| r.name.to_string()).collect();
        assert_display_snapshot!(rule_names.join("\n"));
    }

    /// Ensure we stort the resulting violations by where they occur in the file.
    #[test]
    fn test_check_rules_orderin() {
        let sql = r#"
  ALTER TABLE "table_name" RENAME COLUMN "column_name" TO "new_column_name";
  CREATE INDEX "field_name_idx" ON "table_name" ("field_name");
  "#;

        let res = check_sql(sql, &[RuleViolationKind::PreferRobustStmts], None, false)
            .expect("valid parsing of SQL");
        let mut prev_span_start = -1;
        for violation in &res {
            assert!(violation.span.start > prev_span_start);
            prev_span_start = violation.span.start;
        }
    }
}
