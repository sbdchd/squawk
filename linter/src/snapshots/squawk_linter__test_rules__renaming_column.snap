---
source: linter/src/lib.rs
expression: "check_sql(sql, &[])"
---
Ok(
    [
        RuleViolation {
            kind: RenamingColumn,
            span: Span {
                start: 0,
                len: Some(
                    74,
                ),
            },
            messages: [
                Note(
                    "Renaming a column may break existing clients.",
                ),
            ],
        },
    ],
)
